src/openms/include/OpenMS/DATASTRUCTURES/MatchedIterator.h

```diff
@@ -0,0 +1,326 @@
+                               float tolerance)
+        : ref_begin_(ref_begin), ref_end_(ref_end), tgt_begin_(tgt_begin), tgt_end_(tgt_end), it_ref_(ref_begin), it_tgt_(tgt_begin), tol_(tolerance)
+      {
+        if (tgt_begin_ == tgt_end_)
+        { // nothing to iterate over in target (if ref_ were empty, isEnd_() is automatically true)
+          setToEnd_();
+        }
+        advanceTarget_();
+      }
+
+      /**
+          @brief Default CTor; same as MatchedIterator::end()
+      */ 
+      explicit MatchedIterator()
+        : ref_begin_(), ref_end_(), tgt_begin_(), tgt_end_(), it_ref_(), it_tgt_(), tol_()
+      {
+      }
+
+      /// Copy CTor (default)
+      MatchedIterator(const MatchedIterator& rhs) = default;
+      
+      /// assignment operator (deleted, since we use references to containers)
+      MatchedIterator& operator=(const MatchedIterator& rhs) = delete;
+
+      bool operator==(const MatchedIterator& rhs) const
+      {
+        if (this == &rhs) return true;
+
+        if (isEnd_() || rhs.isEnd_())
+        {
+          return isEnd_() == rhs.isEnd_();
+        }
+
+        return (it_ref_ == rhs.it_ref_ &&
+                it_tgt_ == rhs.it_tgt_ &&
+                ref_begin_ == rhs.ref_begin_ &&
+                ref_end_ == rhs.ref_end_ &&
+                tgt_begin_ == rhs.tgt_begin_ &&
+                tgt_end_ == rhs.tgt_end_);
+      }
+      bool operator!=(const MatchedIterator& rhs) const
+      {
+        return !(*this == rhs);
+      }
+
+      /// dereference current target element
+      template< bool _CONST = CONST >
+      typename std::enable_if< _CONST, reference >::type operator*() const
+      {
+        return *it_tgt_;
+      }
+      template< bool _CONST = CONST >
+      typename std::enable_if< !_CONST, reference >::type operator*()
+      {
+        return *it_tgt_;
+      }
+
+      /// pointer to current target element
+      template< bool _CONST = CONST >
+      typename std::enable_if< _CONST, pointer >::type operator->() const
+      {
+        return &(*it_tgt_);
+      }
+      template< bool _CONST = CONST >
+      typename std::enable_if< !_CONST, pointer >::type operator->()
+      {
+        return &(*it_tgt_);
+      }
+
+      /// current element in reference container
+      const value_type& ref() const
+      {
+        return *it_ref_;
+      }
+
+      /// index into reference container
+      size_t refIdx() const
+      {
+        return it_ref_ - ref_begin_;
+      }
+      
+      /// index into target container
+      size_t tgtIdx() const
+      {
+        return it_tgt_ - tgt_begin_;
+      }
+
+      /**
+        @brief Advances to the next valid pair
+
+        @exception Exception::InvalidIterator If iterator is already at end
+      */
+      MatchedIterator& operator++()
+      {
+        // are we at end already? --> wrong usage
+        if (isEnd_()) throw Exception::InvalidIterator(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
```

Comment:
this should never happen and may introduce an unnecessary check. 

```suggestion
        OPENMS_PRECONDITION(!isEnd_(), "Try to advance beyond end iterator");
```